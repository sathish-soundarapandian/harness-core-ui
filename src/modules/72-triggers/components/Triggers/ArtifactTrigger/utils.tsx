import type { TriggerBaseType, TriggerArtifactType } from '../TriggerInterface'

export interface ArtifactTriggerValues {
  triggerType: TriggerBaseType
  artifactType: TriggerArtifactType
}
