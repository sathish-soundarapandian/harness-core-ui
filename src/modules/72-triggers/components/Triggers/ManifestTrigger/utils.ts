import type { TriggerBaseType, ManifestType } from '../TriggerInterface'

export interface HelmChartInitialValues {
  triggerType: TriggerBaseType
  manifestType: ManifestType
}
