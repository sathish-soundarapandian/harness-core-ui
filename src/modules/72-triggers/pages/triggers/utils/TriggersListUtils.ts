/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
// temporary mock data
import { parse } from 'yaml'
import type { ConnectorInfoDTO } from 'services/cd-ng'
import type { AddDrawerMapInterface, CategoryInterface } from '@common/components/AddDrawer/AddDrawer'
import type { StringKeys, UseStringsReturn } from 'framework/strings'
import {
  manifestTypeIcons,
  ManifestDataType,
  manifestTypeLabels
} from '@pipeline/components/ManifestSelection/Manifesthelper'
import {
  ArtifactIconByType,
  ArtifactTitleIdByType,
  ENABLED_ARTIFACT_TYPES
} from '@pipeline/components/ArtifactsSelection/ArtifactHelper'
import type { TriggerCatalogItem } from 'services/pipeline-ng'
import { TriggerTypes, AWS_CODECOMMIT, AwsCodeCommit } from './TriggersWizardPageUtils'

export const GitSourceProviders: Record<
  string,
  { value: ConnectorInfoDTO['type'] | 'AwsCodeCommit' | 'Custom'; iconName: IconName }
> = {
  GITHUB: { value: 'Github', iconName: 'github' },
  GITLAB: { value: 'Gitlab', iconName: 'service-gotlab' },
  BITBUCKET: { value: 'Bitbucket', iconName: 'bitbucket-selected' },
  AZURE_REPO: { value: 'AzureRepo', iconName: 'service-azure' },
  AWS_CODECOMMIT: { value: 'AwsCodeCommit', iconName: 'service-aws-code-deploy' },
  CUSTOM: { value: 'Custom', iconName: 'build' }
}

const TriggerTypeIcons: Record<'SCHEDULE' | 'NEW_ARTIFACT', IconName> = {
  SCHEDULE: 'trigger-schedule',
  NEW_ARTIFACT: 'new-artifact'
}
export const getTriggerIcon = ({
  type,
  webhookSourceRepo,
  buildType
}: {
  type: string
  webhookSourceRepo?: string // string temporary until backend
  buildType?: string
}): IconName => {
  const updatedWebhookSourceRepo =
    webhookSourceRepo === AwsCodeCommit ? AWS_CODECOMMIT : webhookSourceRepo?.toUpperCase()
  const webhookSourceRepoIconName =
    webhookSourceRepo && updatedWebhookSourceRepo && GitSourceProviders[updatedWebhookSourceRepo]?.iconName
  if (type === TriggerTypes.WEBHOOK && webhookSourceRepoIconName) {
    return webhookSourceRepoIconName as IconName
  } else if (type === TriggerTypes.SCHEDULE) {
    return TriggerTypeIcons.SCHEDULE as IconName
  } else if (type === TriggerTypes.MANIFEST && buildType) {
    if (buildType === ManifestDataType.HelmChart) {
      return manifestTypeIcons.HelmChart
    }
  } else if (type === TriggerTypes.ARTIFACT && buildType) {
    switch (buildType) {
      case ENABLED_ARTIFACT_TYPES.Gcr:
        return ArtifactIconByType.Gcr
      case ENABLED_ARTIFACT_TYPES.Ecr:
        return ArtifactIconByType.Ecr
      case ENABLED_ARTIFACT_TYPES.DockerRegistry:
        return ArtifactIconByType.DockerRegistry
      case ENABLED_ARTIFACT_TYPES.Acr:
        return ArtifactIconByType.Acr
      case ENABLED_ARTIFACT_TYPES.AmazonS3:
        return ArtifactIconByType.AmazonS3
      case ENABLED_ARTIFACT_TYPES.Jenkins:
        return ArtifactIconByType.Jenkins
    }
  }
  return 'yaml-builder-trigger'
}

const triggerDrawerMap = (
  getString: (key: StringKeys) => string,
  isNewService: boolean,
  allowV2Artifacts: boolean | undefined
): AddDrawerMapInterface => ({
  drawerLabel: getString('common.triggersLabel'),
  showAllLabel: getString('triggers.showAllTriggers'),
  categories: [
    {
      categoryLabel: getString('execution.triggerType.WEBHOOK'),
      categoryValue: 'Webhook',
      items: [
        {
          itemLabel: getString('common.repo_provider.githubLabel'),
          value: GitSourceProviders.GITHUB.value,
          iconName: GitSourceProviders.GITHUB.iconName
        },
        {
          itemLabel: getString('common.repo_provider.gitlabLabel'),
          value: GitSourceProviders.GITLAB.value,
          iconName: GitSourceProviders.GITLAB.iconName
        },
        {
          itemLabel: getString('common.repo_provider.bitbucketLabel'),
          value: GitSourceProviders.BITBUCKET.value,
          iconName: GitSourceProviders.BITBUCKET.iconName
        },
        {
          itemLabel: getString('common.repo_provider.azureRepos'),
          value: GitSourceProviders.AZURE_REPO.value,
          iconName: GitSourceProviders.AZURE_REPO.iconName
        },
        {
          itemLabel: getString('common.repo_provider.customLabel'),
          value: GitSourceProviders.CUSTOM.value,
          iconName: GitSourceProviders.CUSTOM.iconName
        }
      ]
    },
    ...(isNewService
      ? [
          {
            categoryLabel: getString('common.comingSoon'),
            categoryValue: 'ArtifactComingSoon'
          }
        ]
      : []),
    {
      categoryLabel: getString('pipeline.artifactTriggerConfigPanel.artifact'),
      categoryValue: 'Artifact',
      items: [
        {
          itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.Gcr]),
          value: ENABLED_ARTIFACT_TYPES.Gcr,
          iconName: ArtifactIconByType.Gcr,
          disabled: isNewService
        },
        {
          itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.Ecr]),
          value: ENABLED_ARTIFACT_TYPES.Ecr,
          iconName: ArtifactIconByType.Ecr,
          disabled: isNewService
        },
        {
          itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.DockerRegistry]),
          value: ENABLED_ARTIFACT_TYPES.DockerRegistry,
          iconName: ArtifactIconByType.DockerRegistry,
          disabled: isNewService
        },
        {
          itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.ArtifactoryRegistry]),
          value: ENABLED_ARTIFACT_TYPES.ArtifactoryRegistry,
          iconName: ArtifactIconByType.ArtifactoryRegistry,
          disabled: isNewService
        },
        {
          itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.Acr]),
          value: ENABLED_ARTIFACT_TYPES.Acr,
          iconName: ArtifactIconByType.Acr,
          disabled: isNewService
        },
        {
          itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.AmazonS3]),
          value: ENABLED_ARTIFACT_TYPES.AmazonS3,
          iconName: ArtifactIconByType.AmazonS3 as IconName,
          disabled: isNewService
        },
        {
          itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.GoogleArtifactRegistry]),
          value: ENABLED_ARTIFACT_TYPES.GoogleArtifactRegistry,
          iconName: ArtifactIconByType.GoogleArtifactRegistry as IconName,
          disabled: isNewService
        },
        {
          itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.CustomArtifact]),
          value: ENABLED_ARTIFACT_TYPES.CustomArtifact,
          iconName: ArtifactIconByType.CustomArtifact as IconName,
          disabled: isNewService
        },
        {
          itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.GithubPackageRegistry]),
          value: ENABLED_ARTIFACT_TYPES.GithubPackageRegistry,
          iconName: ArtifactIconByType.GithubPackageRegistry as IconName,
          disabled: isNewService
        },
        ...(allowV2Artifacts
          ? [
              {
                itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.Jenkins]),
                value: ENABLED_ARTIFACT_TYPES.Jenkins,
                iconName: ArtifactIconByType.Jenkins as IconName,
                disabled: isNewService
              },
              {
                itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.AzureArtifacts]),
                value: ENABLED_ARTIFACT_TYPES.AzureArtifacts,
                iconName: ArtifactIconByType.AzureArtifacts as IconName,
                disabled: isNewService
              },
              {
                itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.Nexus3Registry]),
                value: ENABLED_ARTIFACT_TYPES.Nexus3Registry,
                iconName: ArtifactIconByType.Nexus3Registry as IconName,
                disabled: isNewService
              },
              {
                itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.AmazonMachineImage]),
                value: ENABLED_ARTIFACT_TYPES.AmazonMachineImage,
                iconName: ArtifactIconByType.AmazonMachineImage as IconName,
                disabled: isNewService
              }
            ]
          : [])
      ]
    },
    ...(!isNewService
      ? [
          {
            categoryLabel: getString('common.comingSoon'),
            categoryValue: 'ArtifactComingSoon',
            items: [
              {
                itemLabel: getString(ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.Nexus3Registry]),
                value: ENABLED_ARTIFACT_TYPES.Nexus3Registry,
                iconName: ArtifactIconByType.Nexus3Registry as IconName,
                disabled: true
              }
            ]
          }
        ]
      : []),
    {
      categoryLabel: getString('manifestsText'),
      categoryValue: 'Manifest',
      items: [
        {
          itemLabel: getString(manifestTypeLabels.HelmChart),
          value: ManifestDataType.HelmChart,
          iconName: manifestTypeIcons.HelmChart,
          disabled: isNewService
        }
      ]
    },
    {
      categoryLabel: getString('triggers.scheduledLabel'),
      categoryValue: 'Scheduled',
      items: [
        {
          itemLabel: getString('triggers.cronLabel'),
          value: 'Cron',
          iconName: TriggerTypeIcons.SCHEDULE as IconName
        }
      ]
    }
  ]
})

export const getSourceRepoOptions = (getString: (str: StringKeys) => string): { label: string; value: string }[] => [
  { label: getString('common.repo_provider.githubLabel'), value: GitSourceProviders.GITHUB.value },
  { label: getString('common.repo_provider.gitlabLabel'), value: GitSourceProviders.GITLAB.value },
  { label: getString('common.repo_provider.bitbucketLabel'), value: GitSourceProviders.BITBUCKET.value },
  { label: getString('common.repo_provider.azureRepos'), value: GitSourceProviders.AZURE_REPO.value },
  { label: getString('common.repo_provider.customLabel'), value: GitSourceProviders.CUSTOM.value }
]

export const getCategoryItems = (
  getString: (key: StringKeys) => string,
  isNewService: boolean,
  allowV2Artifacts: boolean | undefined
): AddDrawerMapInterface => triggerDrawerMap(getString, isNewService, allowV2Artifacts)

export interface ItemInterface {
  itemLabel: string
  iconName: IconName
  value: string
  visible?: boolean
  disabled?: boolean
  categoryValue?: string
}

export interface TriggerDataInterface {
  triggerType: string
  sourceRepo?: string
  manifestType?: string
  artifactType?: string
  scheduleType?: string
}

export const getEnabledStatusTriggerValues = ({
  data,
  enabled,
  getString
}: {
  data: any
  enabled: boolean
  getString: (key: StringKeys) => string
}): { values?: any; error?: string } => {
  try {
    const triggerResponseJson = parse(data?.yaml || '')
    triggerResponseJson.trigger.enabled = enabled
    return { values: triggerResponseJson.trigger }
  } catch (e) {
    return { error: getString('triggers.cannotParseTriggersData') }
  }
}
const TriggerStatus = {
  FAILED: 'FAILED',
  UNKNOWN: 'UNKNOWN',
  ERROR: 'ERROR',
  TIMEOUT: 'TIMEOUT',
  UNAVAILABLE: 'UNAVAILABLE',
  SUCCESS: 'SUCCESS'
}

export const errorStatusList = [
  TriggerStatus.FAILED,
  TriggerStatus.UNKNOWN,
  TriggerStatus.ERROR,
  TriggerStatus.TIMEOUT,
  TriggerStatus.UNAVAILABLE
]

const TriggerCategoryToLabelMap: Record<Required<TriggerCatalogItem>['category'], StringKeys> = {
  Webhook: 'execution.triggerType.WEBHOOK',
  Artifact: 'pipeline.artifactTriggerConfigPanel.artifact',
  Manifest: 'manifestsText',
  Scheduled: 'triggers.scheduledLabel'
}

type TriggerCatalogType = Required<TriggerCatalogItem>['triggerCatalogType'][number]

const TriggerCatalogTypeToLabelMap: Record<TriggerCatalogType, StringKeys> = {
  Github: 'common.repo_provider.githubLabel',
  Gitlab: 'common.repo_provider.gitlabLabel',
  Bitbucket: 'common.repo_provider.bitbucketLabel',
  AwsCodeCommit: 'common.repo_provider.awscodecommit',
  Custom: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.CustomArtifact],
  Gcr: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.Gcr],
  Ecr: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.Ecr],
  DockerRegistry: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.DockerRegistry],
  ArtifactoryRegistry: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.ArtifactoryRegistry],
  Acr: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.Acr],
  AmazonS3: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.AmazonS3],
  GoogleArtifactRegistry: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.GoogleArtifactRegistry],
  CustomArtifact: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.CustomArtifact],
  GithubPackageRegistry: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.GithubPackageRegistry],
  Nexus2Registry: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.Nexus2Registry],
  Nexus3Registry: ArtifactTitleIdByType[ENABLED_ARTIFACT_TYPES.Nexus3Registry],
  HelmChart: manifestTypeLabels.HelmChart,
  Cron: 'triggers.cronLabel'
}

const TriggerCatalogTypeToIconMap: Record<TriggerCatalogType, IconName> = {
  Github: GitSourceProviders.GITHUB.iconName,
  Gitlab: GitSourceProviders.GITLAB.iconName,
  Bitbucket: GitSourceProviders.BITBUCKET.iconName,
  AwsCodeCommit: GitSourceProviders.AWS_CODECOMMIT.iconName,
  Custom: GitSourceProviders.CUSTOM.iconName,
  Gcr: ArtifactIconByType.Gcr,
  Ecr: ArtifactIconByType.Ecr,
  DockerRegistry: ArtifactIconByType.DockerRegistry,
  ArtifactoryRegistry: ArtifactIconByType.ArtifactoryRegistry,
  Acr: ArtifactIconByType.Acr,
  AmazonS3: ArtifactIconByType.AmazonS3,
  GoogleArtifactRegistry: ArtifactIconByType.GoogleArtifactRegistry,
  GithubPackageRegistry: ArtifactIconByType.GithubPackageRegistry,
  CustomArtifact: ArtifactIconByType.CustomArtifact,
  Nexus2Registry: ArtifactIconByType.Nexus2Registry,
  Nexus3Registry: ArtifactIconByType.Nexus3Registry,
  HelmChart: manifestTypeIcons.HelmChart,
  Cron: TriggerTypeIcons.SCHEDULE
}

export const getTriggerCategoryDrawerMapFromTriggerCatalogItem = (
  getString: UseStringsReturn['getString'],
  triggerCatalogItems: TriggerCatalogItem[]
): AddDrawerMapInterface => {
  const categories: CategoryInterface[] = triggerCatalogItems.map(catalog => {
    const { category, triggerCatalogType } = catalog
    return {
      categoryLabel: getString(TriggerCategoryToLabelMap[category]),
      categoryValue: category,
      items: triggerCatalogType.map(item => ({
        itemLabel: getString(TriggerCatalogTypeToLabelMap[item]),
        value: item,
        iconName: TriggerCatalogTypeToIconMap[item]
      }))
    }
  })

  return {
    drawerLabel: getString('common.triggersLabel'),
    showAllLabel: getString('triggers.showAllTriggers'),
    categories
  }
}
