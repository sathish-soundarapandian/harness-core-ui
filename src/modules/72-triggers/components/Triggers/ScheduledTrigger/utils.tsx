/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { object, ObjectSchema, string } from 'yup'
import { isEmpty } from 'lodash-es'
import { illegalIdentifiers, regexIdentifier } from '@common/utils/StringUtils'
import { isCronValid } from '@triggers/components/steps/SchedulePanel/components/utils'
import type { StringKeys, UseStringsReturn } from 'framework/strings'
import type { NGTriggerSourceV2, PipelineInfoConfig } from 'services/pipeline-ng'
import type { PanelInterface } from '@triggers/components/TabWizard/TabWizard'
import type { ScheduleType, TriggerBaseType } from '../TriggerInterface'

export interface ScheduledInitialValuesInterface {
  triggerType: TriggerBaseType
  scheduleType: ScheduleType
}

export interface FlatInitialValuesInterface {
  triggerType: NGTriggerSourceV2['type']
  identifier?: string
  tags?: {
    [key: string]: string
  }
  pipeline?: string | PipelineInfoConfig
  originalPipeline?: PipelineInfoConfig
  resolvedPipeline?: PipelineInfoConfig
  inputSetTemplateYamlObj?: {
    pipeline: PipelineInfoConfig | Record<string, never>
  }
  name?: string
  selectedScheduleTab?: string
  pipelineBranchName?: string
  inputSetRefs?: string[]
}

export interface FlatOnEditValuesInterface {
  name?: string
  identifier?: string
  // targetIdentifier: string
  description?: string
  tags?: {
    [key: string]: string
  }
  pipeline?: PipelineInfoConfig
  triggerType: NGTriggerSourceV2['type']
  manifestType?: string
  artifactType?: string
  originalPipeline?: PipelineInfoConfig
  resolvedPipeline?: PipelineInfoConfig
  selectedScheduleTab?: string
  minutes?: string
  expression?: string
  stageId?: string
  inputSetTemplateYamlObj?: {
    pipeline: PipelineInfoConfig | Record<string, never>
  }
  versionValue?: string
  versionOperator?: string
  buildValue?: string
  buildOperator?: string
  pipelineBranchName?: string
  inputSetRefs?: string[]
}

export interface FlatValidScheduleFormikValuesInterface {
  name: string
  identifier: string
  description?: string
  tags?: {
    [key: string]: string
  }
  target?: string
  targetIdentifier?: string
  pipeline: PipelineInfoConfig
  resolvedPipeline?: PipelineInfoConfig
  sourceRepo: string
  triggerType: NGTriggerSourceV2['type']
  expression: string
  pipelineBranchName?: string
  inputSetRefs?: string[]
}

const isIdentifierIllegal = (identifier: string): boolean =>
  regexIdentifier.test(identifier) && illegalIdentifiers.includes(identifier)

const checkValidOverview = ({ formikValues }: { formikValues: { [key: string]: any } }): boolean =>
  isIdentifierIllegal(formikValues?.identifier) ? false : true

const checkValidCronExpression = ({ formikValues }: { formikValues: { [key: string]: any } }): boolean =>
  isCronValid(formikValues?.expression || '')

const checkValidPipelineInput = ({ formikErrors }: { formikErrors: { [key: string]: any } }): boolean => {
  if (!isEmpty(formikErrors?.pipeline) || !isEmpty(formikErrors?.stages)) {
    return false
  }
  return true
}

export const getPanels = (getString: UseStringsReturn['getString']): PanelInterface[] | [] => {
  return [
    {
      id: 'Trigger Overview',
      tabTitle: getString('overview'),
      checkValidPanel: checkValidOverview,
      requiredFields: ['name', 'identifier'] // conditional required validations checkValidTriggerConfiguration
    },
    {
      id: 'Schedule',
      tabTitle: getString('common.schedule'),
      checkValidPanel: checkValidCronExpression,
      requiredFields: ['expression']
    },
    {
      id: 'Pipeline Input',
      tabTitle: getString('triggers.pipelineInputLabel'),
      checkValidPanel: checkValidPipelineInput
    }
  ]
}

export const getValidationSchema = (
  getString: (key: StringKeys, params?: any) => string
): ObjectSchema<Record<string, any> | undefined> => {
  return object().shape({
    name: string().trim().required(getString('triggers.validation.triggerName')),
    identifier: string().when('name', {
      is: val => val?.length,
      then: string()
        .required(getString('validation.identifierRequired'))
        .matches(regexIdentifier, getString('validation.validIdRegex'))
        .notOneOf(illegalIdentifiers)
    }),
    expression: string().test(
      getString('triggers.validation.cronExpression'),
      getString('triggers.validation.cronExpression'),
      function (expression) {
        return isCronValid(expression || '')
      }
    )
  })
}
