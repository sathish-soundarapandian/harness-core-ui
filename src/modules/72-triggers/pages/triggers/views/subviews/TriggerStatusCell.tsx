/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useCallback, useState } from 'react'
import { defaultTo } from 'lodash-es'
import { Color } from '@harness/design-system'
import { Button, ButtonVariation, IconName, IconProps, Layout, Text } from '@harness/uicore'
import { UseStringsReturn, useStrings } from 'framework/strings'
import type { TriggerStatus } from 'services/pipeline-ng'
import { TriggerStatusEnum } from '../../utils/TriggersListUtils'
import TriggerStatusErrorModal from './TriggerStatusErrorModal/TriggerStatusErrorModal'
import css from '../TriggersListSection.module.scss'

export interface TriggerStatusProps {
  triggerStatus: TriggerStatus
}

const getTriggerStatusMessagesMap = (getString: UseStringsReturn['getString']): Map<TriggerStatus['status'], string> =>
  new Map([
    ['SUCCESS', getString('success')],
    ['FAILED', getString('failed')],
    ['UNKNOWN', getString('common.unknown')]
  ])

const triggerStatusIconMap = new Map<TriggerStatus['status'], { icon?: IconName; iconProps?: Partial<IconProps> }>([
  ['SUCCESS', { icon: 'full-circle' as IconName, iconProps: { size: 6, color: Color.GREEN_500 } }],
  ['FAILED', { icon: 'warning-sign' as IconName, iconProps: { size: 12, color: Color.RED_500 } }],
  ['UNKNOWN', {}]
])

export default function TriggerStatusCell({ triggerStatus }: TriggerStatusProps): React.ReactElement {
  const { status, detailMessages } = triggerStatus
  const { getString } = useStrings()
  const [modalOpen, setModalOpen] = useState(false)

  const triggerStatusMessagesMap = getTriggerStatusMessagesMap(getString)
  const statusMsg = defaultTo(triggerStatusMessagesMap.get(status), '')
  const statusIcon = defaultTo(triggerStatusIconMap.get(status), {})

  const handleErrorDetailsClick = (e: React.SyntheticEvent): void => {
    e.stopPropagation()
    setModalOpen(true)
  }

  const renderTooltip = useCallback(() => {
    if (status === TriggerStatusEnum.SUCCESS) return

    return (
      <Layout.Vertical font={{ size: 'small' }} spacing="small" padding="small">
        {detailMessages?.map((error, idx) => (
          <Text key={idx} font={{ size: 'small' }} color={Color.WHITE} lineClamp={2} tooltipProps={{ disabled: true }}>
            {error}
          </Text>
        ))}
        <Button
          minimal
          text={getString('common.viewErrorDetails')}
          onClick={handleErrorDetailsClick}
          variation={ButtonVariation.LINK}
        />
      </Layout.Vertical>
    )
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [status, detailMessages, triggerStatus])

  return (
    <div onClick={e => e.stopPropagation()}>
      <Text
        inline
        icon={statusIcon.icon}
        iconProps={statusIcon.iconProps}
        tooltip={renderTooltip()}
        tooltipProps={{
          isDark: true,
          position: 'bottom',
          popoverClassName: css.tooltip,
          className: css.tooltipWrapper
        }}
      >
        {statusMsg}
      </Text>
      <TriggerStatusErrorModal
        closeDialog={() => setModalOpen(false)}
        isOpen={modalOpen}
        triggerStatus={triggerStatus}
      />
    </div>
  )
}
