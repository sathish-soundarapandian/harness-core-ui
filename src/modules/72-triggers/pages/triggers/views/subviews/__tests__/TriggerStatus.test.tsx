/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { fireEvent, render, waitFor } from '@testing-library/react'
import type { TriggerStatus } from 'services/pipeline-ng'
import { TestWrapper } from '@common/utils/testUtils'
import TriggerStatusCell from '../TriggerStatusCell'

describe('Test case for trigger status rendering and behavior', () => {
  test('should render SUCCESS status properly', async () => {
    const triggerStatusMock: TriggerStatus = { status: 'SUCCESS', detailMessages: [] }
    const { getByText } = render(
      <TestWrapper>
        <TriggerStatusCell triggerStatus={triggerStatusMock} />
      </TestWrapper>
    )

    expect(getByText('success')).toBeDefined()
  })

  test('should render FAILED status properly', async () => {
    const triggerStatusMock: TriggerStatus = { status: 'FAILED', detailMessages: [] }
    const { getByText } = render(
      <TestWrapper>
        <TriggerStatusCell triggerStatus={triggerStatusMock} />
      </TestWrapper>
    )

    expect(getByText('failed')).toBeDefined()
  })

  test('should render UNKNOWN status properly', async () => {
    const triggerStatusMock: TriggerStatus = { status: 'UNKNOWN', detailMessages: [] }
    const { getByText } = render(
      <TestWrapper>
        <TriggerStatusCell triggerStatus={triggerStatusMock} />
      </TestWrapper>
    )

    expect(getByText('common.unknown')).toBeDefined()
  })

  test('should render without crash if data is not provided', async () => {
    const triggerStatusMock: TriggerStatus = {}
    const { container } = render(
      <TestWrapper>
        <TriggerStatusCell triggerStatus={triggerStatusMock} />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })

  test('should render without crash if errors are not provided', async () => {
    const triggerStatusMock: TriggerStatus = { status: 'FAILED' }
    const { findByText, getByText } = render(
      <TestWrapper>
        <TriggerStatusCell triggerStatus={triggerStatusMock} />
      </TestWrapper>
    )

    fireEvent.mouseEnter(getByText('failed'))
    expect(await findByText('common.viewErrorDetails')).toBeDefined()
  })

  test('should show popover on mouse over on FAILED status', async () => {
    const triggerStatusMock: TriggerStatus = { status: 'FAILED', detailMessages: ['error 1'] }
    const { getByText, findByText } = render(
      <TestWrapper>
        <TriggerStatusCell triggerStatus={triggerStatusMock} />
      </TestWrapper>
    )

    fireEvent.mouseEnter(getByText('failed'))

    expect(await findByText('error 1')).toBeDefined()
    expect(await findByText('common.viewErrorDetails')).toBeDefined()
  })

  test('shoud open (and close) modal with error messages', async () => {
    const detailMessages = ['error 1', 'error 2']
    const triggerStatusMock: TriggerStatus = { status: 'FAILED', detailMessages }
    const { getByText, findByText, findByRole, findByTestId, queryByTestId, baseElement } = render(
      <TestWrapper>
        <TriggerStatusCell triggerStatus={triggerStatusMock} />
      </TestWrapper>
    )

    // open tooltip
    fireEvent.mouseEnter(getByText('failed'))

    // open modal
    const btn = await findByRole('button')
    expect(btn).toBeDefined()
    fireEvent.click(btn)

    const title = await findByText('errorDetails')
    expect(title).toBeDefined()

    const preEl = await findByTestId('trigger-status-errors')
    expect(preEl).toBeDefined()
    const expectedText = JSON.stringify(detailMessages, null, 1)
    expect(preEl.innerHTML).toMatch(expectedText)

    // close modal
    const closeBtn = baseElement.querySelector('button[aria-label="Close"]')
    fireEvent.click(closeBtn!)

    await waitFor(() => expect(queryByTestId('modaldialog-body')).toBeNull())
  })
})
