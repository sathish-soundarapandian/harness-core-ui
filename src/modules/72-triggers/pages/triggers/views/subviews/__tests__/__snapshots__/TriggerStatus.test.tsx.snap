// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for trigger status rendering and behavior should render without crash if data is not provided 1`] = `
<div>
  <div>
    <span
      class="bp3-popover-wrapper tooltipWrapper"
    >
      <span
        class="bp3-popover-target Text--targetClass"
      >
        <span
          class="StyledProps--font StyledProps--main StyledProps--inline"
          tabindex="0"
        >
          
        </span>
      </span>
    </span>
  </div>
</div>
`;
